### PR TITLE
Estebanreyl/ready for production

### PR DESCRIPTION
**Pull request by estebanreyl**
- Updated Create registry, delete registry, delete image and delete repository to use ample code sharing.  Updated token acquisition
- Fixed delete registry
- Fixed delete repository and delete image's limitations when acting on registries without admin enabled
- Ample code styling

As a note, this PR is meant to prepare for a PR to Microsoft/vscode docker, as a result, I will soon open a PR request to upstream with everything in here minus build task nodes.
@sajayantony , @northtyphoon any suggestions on how best to proceed would be appreciated, should we merge this in directly for further internal development and open a new PR for upstream by tomorrow?

### CODE REVIEW

File: utils/Azure/acrTools.ts [hunk 1]
```diff
@@ -1,273 +1,164 @@
...
         http_method: http_method,
         url: url
     }
     if (http_method === 'delete') {
-        await request.delete(opt);
-        vscode.window.showInformationMessage('Successfully deleted item');
+        try {
+            await request.delete(opt);
+        } catch (error) {
+            console.log(error);
```
[the remark this answers is not in the document]
Comment (estebanreyl) [reply utils/Azure/acrTools.ts]: Oh we can ignore that I forgot to remove that, it was for testing

File: utils/Azure/acrTools.ts [hunk 2]
```diff
@@ -1,273 +1,164 @@
...
+    const { acrRefreshToken, acrAccessToken } = await getDockerCompatibleTokensFromRegistry(registry, "registry:catalog:*");
+    await request.get('https://' + registry.loginServer + '/v2/_catalog', {
+        auth: {
+            bearer: acrAccessToken
+        }
+    }, (err, httpResponse, body) => {
+        if (body.length > 0) {
+            const repositories = JSON.parse(body).repositories;
+            for (let tempRepo of repositories) {
+                repo = new Repository(registry, tempRepo, acrAccessToken, acrRefreshToken);
```
[the remark this answers is not in the document]
Comment (estebanreyl) [reply utils/Azure/acrTools.ts]: That makes sense

File: utils/Azure/acrTools.ts [hunk 3]
```diff
@@ -1,273 +1,164 @@
...
-            if (err) { return []; }
-
-            if (body.length > 0) {
-                tags = JSON.parse(body).tags;
+/** List repositories on a given Registry. */
+export async function getACRRepositories(registry: Registry): Promise<Repository[]> {
+    const allRepos: Repository[] = [];
+    let repo: Repository;
+    const { acrRefreshToken, acrAccessToken } = await getDockerCompatibleTokensFromRegistry(registry, "registry:catalog:*");
+    await request.get('https://' + registry.loginServer + '/v2/_catalog', {
```
[the remark this answers is not in the document]
Comment (estebanreyl) [reply utils/Azure/acrTools.ts]: Potentially, Ill review that

File: commands/azureCommands/delete-azure-image.ts [hunk 4]
```diff
@@ -10,54 +11,40 @@ import { AzureUtilityManager } from '../../utils/azureUtilityManager';
...
     }
 
-    let creds = await acrTools.loginCredentials(subscription, registry);
-    username = creds.username;
-    password = creds.password;
-    let path = `/v2/_acr/${repoName}/tags/${tag}`;
-    await acrTools.sendRequestToRegistry('delete', registry.loginServer, path, username, password); //official call to delete the image
+    const shouldDelete = await quickPicks.confirmUserIntent('Are you sure you want to delete this image? Enter Yes to continue: ');
+    if (shouldDelete) {
+        const { acrAccessToken } = await acrTools.getDockerCompatibleTokensFromRegistry(registry, `repository:${repoName}:*`);
```
[the remark this answers is not in the document]
Comment (estebanreyl) [reply commands/azureCommands/delete-azure-image.ts]: Yes, I was having trouble with naming conventions here ideally it should do the same as acquireACRAccessToken just having different requirements but typescript doesn't support function overloading, is there a best practice naming convention in this case? Say acquireACRAccessTokenFromRegistry?

File: utils/Azure/acrTools.ts [hunk 5]
```diff
@@ -1,273 +1,164 @@
...
-                accessTokenACR = JSON.parse(body).access_token;
-            } else {
-                return;
-            }
-        });
-        if (refreshTokenACR && accessTokenACR) {
-            return { 'refreshToken': refreshTokenACR, 'accessToken': accessTokenACR };
+//Registry item management
+/** List images under a specific Repository */
+export async function getACRImages(element: Repository): Promise<AzureImage[]> {
```
[the remark this answers is not in the document]
Comment (estebanreyl) [reply utils/Azure/acrTools.ts]: That seems reasonable, I'll only keep ACR for the tokens as those make things clearer given the aad tokens.

File: utils/Azure/acrTools.ts [hunk 6]
```diff
@@ -1,273 +1,164 @@
...
+export async function sendRequestToRegistry(http_method: string, login_server: string, path: string, password: string): Promise<any> {
     let url: string = `https://${login_server}${path}`;
-    let header = getAuthorizationHeader(username, password);
+    let header = 'Bearer ' + password;
     let opt = {
         headers: { 'Authorization': header },
         http_method: http_method,
         url: url
     }
     if (http_method === 'delete') {
```
[the remark this answers is not in the document]
Comment (estebanreyl) [reply utils/Azure/acrTools.ts]: As its implemented yes, could probably modify it without much work however

File: commands/azureCommands/delete-azure-image.ts [hunk 7]
```diff
@@ -10,54 +11,40 @@ import { AzureUtilityManager } from '../../utils/azureUtilityManager';
...
-    answer = answer.toLowerCase();
-    if (answer !== 'yes') { return; }
-
-    if (context) {
-        repoName = context.label;
-        subscription = context.subscription;
+    } else {
         registry = context.registry;
-        let wholeName = repoName.split(':');
+        let wholeName = context.label.split(':');
```
Comment (rsamai) [inline commands/azureCommands/delete-azure-image.ts]: nit: add type to make it more clear

File: commands/azureCommands/delete-azure-registry.ts [hunk 8]
```diff
@@ -19,24 +20,17 @@ export async function deleteAzureRegistry(context?: AzureRegistryNode): Promise<
     } else {
         registry = await quickPickACRRegistry();
     }
+    const shouldDelete = await confirmUserIntent('Are you sure you want to delete this registry and its associated images? Enter yes to continue: ');
+    if (shouldDelete) {
+        let subscription: SubscriptionModels.Subscription = acrTools.getRegistrySubscription(registry);
+        let resourceGroup: string = acrTools.getResourceGroupName(registry);
+        const client = AzureUtilityManager.getInstance().getContainerRegistryManagementClient(subscription);
```
Comment (rsamai) [inline commands/azureCommands/delete-azure-registry.ts]: Should we make a quick function for this?
getAzureUtilityManager_managementClient(subscription) or something if we're going about simplifying things  + we use this in a ton of places

File: commands/azureCommands/delete-repository.ts [hunk 9]
```diff
@@ -23,29 +24,20 @@ export async function deleteRepository(context?: AzureRepositoryNode): Promise<v
...
+    const shouldDelete = await confirmUserIntent('Are you sure you want to delete this repository and its associated images? Enter yes to continue: ');
+    if (shouldDelete) {
+        const { acrAccessToken } = await acrTools.getDockerCompatibleTokensFromRegistry(registry, `repository:${repoName}:*`);
+        const path = `/v2/_acr/${repoName}/repository`;
+        await acrTools.sendRequestToRegistry('delete', registry.loginServer, path, acrAccessToken);
+        vscode.window.showInformationMessage(`Successfully deleted repository ${Repository}`);
+    } else {
+        throw new UserCancelledError();
+    }
     reportTelemetry();
```
Comment (rsamai) [inline commands/azureCommands/delete-repository.ts]: You wrote out this function in both delete-repository and delete-azure-image. Put in central loc?

File: commands/azureCommands/delete-azure-registry.ts [hunk 10]
```diff
@@ -19,24 +20,17 @@ export async function deleteAzureRegistry(context?: AzureRegistryNode): Promise<
...
-    answer = answer.toLowerCase();
-    if (answer !== 'yes') { return; }
-
-    let subscription: SubscriptionModels.Subscription = acrTools.getRegistrySubscription(registry);
-    let resourceGroup: string = acrTools.getResourceGroupName(registry);
-    const client = AzureUtilityManager.getInstance().getContainerRegistryManagementClient(subscription);
-
-    await client.registries.beginDeleteMethod(resourceGroup, registry.name);
-    vscode.window.showInformationMessage('Successfully deleted registry ' + registry.name);
     telemetryReport();
```
Comment (rsamai) [inline commands/azureCommands/delete-azure-registry.ts]: Doesn't this do the same thing as reportTelemetry() ?

File: commands/utils/quick-pick-azure.ts [hunk 11]
```diff
@@ -40,19 +34,155 @@ export async function quickPickACRRepository(registry: Registry): Promise<Reposi
...
+        locationNames.push(loc.displayName);
+    }
+
+    locationNames.sort((loc1: string, loc2: string): number => {
+        return loc1.localeCompare(loc2);
+    });
+
+    let location: string = await vscode.window.showQuickPick(locationNames, {
+        'canPickMany': false,
+        'placeHolder': 'Choose a Location to use'
```
Comment (rsamai) [inline commands/utils/quick-pick-azure.ts]: nit: "Choose a Location" to be consistent with other placeholders

File: commands/utils/quick-pick-azure.ts [hunk 12]
```diff
@@ -40,19 +34,155 @@ export async function quickPickACRRepository(registry: Registry): Promise<Reposi
...
+        ignoreFocusOut: true,
+        placeHolder: 'No',
+        value: 'No',
+        prompt: yesOrNoPrompt
+    };
+    let answer = await vscode.window.showInputBox(opt);
+    if (!answer) { throw new UserCancelledError(); }
+
+    answer = answer.toLowerCase();
+    return answer === 'yes';
```
Comment (rsamai) [inline commands/utils/quick-pick-azure.ts]: nit: make sure it's capital "Yes"

File: commands/utils/quick-pick-azure.ts [hunk 13]
```diff
@@ -40,19 +34,155 @@ export async function quickPickACRRepository(registry: Registry): Promise<Reposi
...
+        registries = await AzureUtilityManager.getInstance().getRegistries(); // Reload
+    }
+
     const registry = registries.find((currentReg): boolean => { return desired === currentReg.name });
     return registry;
 }
+
+export async function quickPickSKU(): Promise<string> {
+    let sku: string;
+    sku = await vscode.window.showQuickPick(skus, { 'canPickMany': false, 'placeHolder': 'Choose a SKU' });
```
Comment (rsamai) [inline commands/utils/quick-pick-azure.ts]: "Choose a SKU to use" to be consistent with other placeholders

File: commands/azureCommands/create-registry.ts [hunk 14]
```diff
@@ -1,217 +1,61 @@
...
-    if (sku === undefined) { throw new Error('User exit'); }
-
-    return sku;
 }
 
+/** Acquires a new registry name from a user, validating that the name is unique */
 async function acquireRegistryName(client: ContainerRegistryManagementClient): Promise<string> {
     let opt: vscode.InputBoxOptions = {
         ignoreFocusOut: false,
         prompt: 'Registry name? '
```
Comment (rsamai) [inline commands/azureCommands/create-registry.ts]: New Registry name? to be consistent with "New Resource Group name?"

File: commands/utils/quick-pick-azure.ts [hunk 15]
```diff
@@ -40,19 +34,155 @@ export async function quickPickACRRepository(registry: Registry): Promise<Reposi
...
+    return answer === 'yes';
+}
+
+/*Creates a new resource group within the current subscription */
+async function createNewResourceGroup(loc: string, subscription?: Subscription): Promise<ResourceGroup> {
+    const resourceGroupClient = AzureUtilityManager.getInstance().getResourceManagementClient(subscription);
+
+    let opt: vscode.InputBoxOptions = {
+        ignoreFocusOut: false,
+        prompt: 'New resource group name?'
```
Comment (rsamai) [inline commands/utils/quick-pick-azure.ts]: capitalize Resource Group. The capitalization kinda goes back and forth between prompts. So why don't you stick with capitalizing the thing in question (ie Location, Registry, etc)

File: commands/utils/quick-pick-azure.ts (hunk 12 above)
Comment (estebanreyl) [reply commands/utils/quick-pick-azure.ts]: any yes should work, this is intended

File: commands/utils/quick-pick-azure.ts (hunk 15 above)
Comment (estebanreyl) [reply commands/utils/quick-pick-azure.ts]: Seems reasonable, I often do this unknowingly.

**Comment by estebanreyl** [comment]
I think this has all been updated now

**Pull request merged**